include/gsl/gsl-lite.hpp

```diff
@@ -1367,31 +1453,70 @@ class not_null
     gsl_api gsl_constexpr   not_null( not_null const & other ) : ptr_ ( other.ptr_  ) {}
     gsl_api                 not_null & operator=( not_null const & other ) { ptr_ = other.ptr_; return *this; }
 # if gsl_HAVE( RVALUE_REFERENCE )
-    gsl_api gsl_constexpr   not_null( not_null && other ) : ptr_( std::move( other.get() ) ) {}
-    gsl_api                 not_null & operator=( not_null && other ) { ptr_ = std::move( other.get() ); return *this; }
+    gsl_api gsl_constexpr   not_null( not_null && other ) : ptr_( std::move(other.checked_ptr()) ) {}
+    gsl_api                 not_null & operator=( not_null && other ) { ptr_ = std::move(other.checked_ptr()); return *this; }
 # endif
 #endif
 
+#if gsl_HAVE( RVALUE_REFERENCE )
+    // explicit converting constructor
+    // if type_traits is not available, then we can't distinguish is_convertible and is_constructible, so we use this unconditionally
+
     template< class U
-#if gsl_HAVE( DEFAULT_FUNCTION_TEMPLATE_ARG )
-        gsl_REQUIRES_T(( std::is_convertible<U, T>::value ))
-#endif
+# if gsl_HAVE( DEFAULT_FUNCTION_TEMPLATE_ARG ) && gsl_HAVE( TYPE_TRAITS )
+        gsl_REQUIRES_A((std::is_constructible<T, U>::value && !std::is_convertible<U, T>::value))
+# endif
     >
-    gsl_api gsl_constexpr not_null( not_null<U> const & other )
-    : ptr_( other.get() )
+    gsl_api gsl_constexpr explicit not_null( not_null<U> other )
+    : ptr_( std::move(other.checked_ptr()) )
+    {}
+
+# if gsl_HAVE( DEFAULT_FUNCTION_TEMPLATE_ARG ) && gsl_HAVE( TYPE_TRAITS )
+    // implicit converting constructor if type_traits is available
+    template< class U
+        gsl_REQUIRES_A((std::is_convertible<U, T>::value))
+    >
+    gsl_api gsl_constexpr not_null( not_null<U> other )
+    : ptr_( std::move(other.checked_ptr()) )
+    {}
+# else
+    // there's no implicit converting constructor, so we need to define the assignment operator manually
+    template< class U >
+    gsl_api gsl_constexpr14 not_null<T>& operator=( not_null<U> other)
+    {
+        ptr_ = std::move(other.checked_ptr());
+        return *this;
+    }
+# endif
+#else // a.k.a. #if ! gsl_HAVE( RVALUE_REFERENCE )
+    template< class U >
+    gsl_api gsl_constexpr explicit not_null( const not_null<U>& other )
+    : ptr_( other.checked_ptr() )
     {}
+    
+    template< class U >
+    gsl_api gsl_constexpr14 not_null<T>& operator=( const not_null<U>& other)
+    {
+        ptr_ = other.checked_ptr();
+        return *this;
+    }
+#endif // gsl_HAVE( RVALUE_REFERENCE )
 
     gsl_api gsl_constexpr14 get_result_t get() const
     {
-        // Without cheating and changing ptr_ from the outside, this check is superfluous:
```

Comment:
This comment was wrong (`ptr_` can legally be `nullptr` after the pointer was moved-from) so it's a good thing you took it out.